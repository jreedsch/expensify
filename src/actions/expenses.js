// ACTION GENERATORS FOR STORE EXPENSES ATTRIBUTE

import uuid from 'uuid';

export const addExpense = ({ description = '', note = '', amount = 0, createdAt = 0 } = {}) => ({
  type: 'ADD_EXPENSE',
  expense: {
    id: uuid(), description, note, amount, createdAt
  }
});

export const removeExpense = ({ id } = {}) => {
  return {
    type: 'REMOVE_EXPENSE',
    id
  }
}

export const editExpense = (id, updates) => {
  console.log("IN action creator, id: "+id+", updates: "+JSON.stringify(updates));
  return (
    { type: 'EDIT_EXPENSE',
      id,
      updates
    }
  );
};

//export const editExpense = (id, updates) => ({
//  type: 'EDIT_EXPENSE',
//  id,
//  updates
//});
