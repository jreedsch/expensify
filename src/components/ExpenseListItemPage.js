import React from 'react';
//import { connect } from 'react-redux';
//import { removeExpense } from '../actions/expenses';
import { Link }  from 'react-router-dom';
import moment from 'moment';

//const ExpenseListItemPage = (props) => ( //implicit return
//const ExpenseListItemPage = ({dispatch, description, amount, createdAt, id}) => { //destructured, from caller's spread operator
const ExpenseListItemPage = ({description, amount, createdAt, id}) => { //destructured, from caller's spread operator
//export const ExpenseListItemPage = ({description, amount, createdAt, id}) => { //destructured, from caller's spread operator
  const createdAtMoment = moment(createdAt).format("MMM Do YYYY");
  return (
    <div>
      <p><Link to={`/edit/${id}`}>{description}</Link></p>
      <p>amount: {amount}, createdAt: {createdAtMoment}</p>



    </div>
  );
};

//      <button onClick={(event) => {dispatch(removeExpense({id}))}}>Remove!</button>

//    <p>description: {props.expense.description} amount: {props.expense.amount} createdAt: {props.expense.createdAt}</p>


// just get dispatch from connect
//export default connect()(ExpenseListItemPage);
export default ExpenseListItemPage;
