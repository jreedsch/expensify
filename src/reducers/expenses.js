const expensesReducerDefaultState = [];

//export default (state = expensesReducerDefaultState, action) => {
const expensesReducer = (state = expensesReducerDefaultState, action) => {
  console.log("IN expense reducer, state: "+JSON.stringify(state));
  console.log("IN expense reducer, action: "+JSON.stringify(action));
  switch (action.type) {
    case 'ADD_EXPENSE':
      console.log("IN expenses reducer, add expense: ");
      console.log({...action.expense});
      return [ ...state, action.expense ] //state.concat(action.expense);

    case 'SET_EXPENSES': //wipe state expenses
      console.log("IN expenses reducer SET_EXPENSES, expenses: ");
      console.log({...action.expenses});
      return action.expenses;  //[ ...state, ...action.expenses ]
      //return state;

    case 'LOAD_EXPENSES': //wipe state expenses
      console.log("IN expenses reducer LOAD_EXPENSES, expenses: ");
      console.log({...action.expenses});
      return action.expenses;  //[ ...state, ...action.expenses ]

    case 'REMOVE_EXPENSE':
      console.log("IN expenses reducer REMOVE_EXPENSE, id: "+action.id);
      //return state.filter((expense) => { return expense.id !== action.expense.id});
      //return state.filter(( {id} ) => { return id !== action.id});
      return state.filter(( {id} ) => id !== action.id );

    case 'EDIT_EXPENSE':
      return state.map((expense) => {
          if (expense.id === action.id) {
            console.log("IN expense reducer, expense: "+JSON.stringify(expense));
            return {...expense, ...action.updates}; //override existing attrs with updates object attrs
          } else {
            return expense;
          }
        }
      );
    default:
      return state;
  }
};

export default expensesReducer;
