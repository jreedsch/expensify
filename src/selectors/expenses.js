import moment from 'moment';

// filter views
const getVisibleExpensesX = (expenses, filters) => {
  console.log("IN getVisibleExpenses, expenses: "+expenses);
  console.log("IN getVisibleExpenses, filters: "+filters);
}

export const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
  return expenses.filter((expense) => {
      //console.log("expense: "+JSON.stringify(expense));
      //console.log("startDate: "+startDate);
      //console.log("endDate: "+endDate);
      //console.log("expense.createdAt: "+expense.createdAt);
      //console.log("text: "+text);
      //console.log("expense.description.toLowerCase(): "+expense.description.toLowerCase());
      //const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate; //disregard if not a number
      //const endDateMatch   = typeof endDate !== 'number' || expense.createdAt <= endDate;

      const createdAtMoment = moment(expense.createdAt);

      const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
      const endDateMatch   = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;


      let textMatch        = typeof text !== 'string' || expense.description.toLowerCase().includes(text.toLowerCase());

      //console.log("textMatch: "+textMatch);

      return startDateMatch && endDateMatch && textMatch;
    }
  ).sort((a,b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'amount'){
      return a.amount < b.amount? 1 : -1;
    }
  })
}

//export default getVisibleExpenses;

// sum up all expense amounts
export const getExpenseTotal = (expenses) => {

  //const expenseAmount = expenses.map((expense) => {
  //  return expense.amount;
  //})
  //return expenseAmount.reduce((acc, val) => {
  //  return acc + val;
  //}, 0);

  return expenses
         .map((expense) => expense.amount)
         .reduce((sum,value) => sum + value, 0);
}
