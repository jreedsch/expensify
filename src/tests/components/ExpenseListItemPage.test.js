import React from 'react';
import { shallow } from 'enzyme';
import ExpenseListItemPage from '../../components/ExpenseListItemPage'; //there is no connected version
import expenses from '../fixtures/expenses';

test('should render ExpenseListItemPage with test expense',() =>{
  const expense = expenses[0];
  const wrapper = shallow(<ExpenseListItemPage {...expense}  />);
  expect(wrapper).toMatchSnapshot();
})
