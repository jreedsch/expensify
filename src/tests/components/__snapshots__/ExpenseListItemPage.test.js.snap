// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItemPage with test expense 1`] = `
<div>
  <p>
    <Link
      replace={false}
      to="/edit/1"
    >
      bread
    </Link>
  </p>
  <p>
    amount: 
    395
    , createdAt: 
    Dec 31st 1969
  </p>
</div>
`;
